### PR TITLE
[SecurityBundle] Allow accessing remember_me services via FirewallMap

### PR DESCRIPTION
**Pull request by chalasr**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #3137, #11158 (not sure) |
| License | MIT |
| Doc PR | todo |

This allows to retrieve the `RememberMeServices` service of the current firewall using:

``` php
FirwallMap::getRememberMeServices(Request $request) : RememberMeServicesInterface
```

An alternative could be to create a public alias of the service.
It seems there are use cases, see the ticket and the ones mentioning it (some from FOSUserBundle).

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php
Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php]: What about caching contexts by requests using a `SplObjectStorage`?

**Comment by iltar** [comment]
So does this encourage developers to avoid the security configuration to do it programmatically outside of the firewall? If so, sounds like a bad idea. All authentication related stuff should go through the firewall.  If this is a DX issue, maybe that should be solved instead?

**Comment by chalasr** [comment]
@iltar That's not the goal of this PR, although you're right, it can be a side effect. 
I didn't just create a public alias of this service nor propose a LoginManager as [FOSUB did it](https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Security/LoginManager.php) for avoiding wrong usages by accessing the service of any firewall . The most relevant use case to me is when using third party libraries for authenticating users such as OAuth, where it can be useful to access the remember me services for persistence. ATM I don't see any alternative for fixing this issue while being restrictive on this point. Any clue?

File: src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php]: It could be a good candidate I think!

**Comment by iltar** [comment]
I don't know any alternative so I guess this is the way to go for the case at hand

**Comment by chalasr** [comment]
@iltar Just re-read you.

> to avoid the security configuration

This allows to access the `remember_me` services of the _current_ firewall, involving to have this one secured, so it requires configuration as well as per the classic way (at least setting up a `remember_me`, which needs a form-positioned security factory configured, a user provider, etc.).
So this just allows to be authenticated by requesting a different endpoint than the login one, e.g. from a registration confirmation sent by email, which is IMHO useful.

BTW, Guard already provides a way to authenticate users programatically (which, contrarily to this, doesn't seem to require any configuration) via [`GuardAuthenticatorHandler::authenticateUserAndHandleSuccess()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php).

This PR could give a structured way to solve the related issues, and it is IMHO a nice have.

File: src/Symfony/Bundle/SecurityBundle/Tests/Security/FirewallContextTest.php
Comment (lemoinem) [inline src/Symfony/Bundle/SecurityBundle/Tests/Security/FirewallContextTest.php]: Wouldn't `assertSame` be better in these cases?

**Review by lemoinem — CHANGES_REQUESTED** [review]
Hi,

Thanks for the contribution. If you could take a look at my comment regarding the tests...

If you could also provide a docs PR, I think we would be good to go to Reviewed status.

**Comment by lemoinem** [comment]
Status: Needs work

**Comment by chalasr** [comment]
@lemoinem I updated the test according to your suggestion.

About the doc PR, I would like to have at least a thought from the core team before submitting it. Doc PRs being most often created after merging the code PR, it should not be a problem for marking this as reviewed.

Thank you for your review

**Review by lemoinem — APPROVED** [review]
Looks good to me! :+1:

**Comment by lemoinem** [comment]
Status: Reviewed

File: src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php]: Done in #20196

**Comment by chalasr** [comment]
Not ready as is.
Status: needs work

**Comment by chalasr** [comment]
Closing this as it doesn't look the best way to fix the related tickets.

**Pull request closed without merging**
